stacks-signer/src/chainstate.rs

```diff
@@ -0,0 +1,375 @@
+    /// The signer thinks this sortition's miner is invalid, but already signed one or more blocks for them.
+    InvalidatedAfterFirstBlock,
+    /// The signer thinks this sortition's miner is valid
+    Valid,
+}
+
+/// Captures the Stacks sortition related state for
+///  a successful sortition
+pub struct SortitionState {
+    /// The miner's pub key hash
+    pub miner_pkh: Hash160,
+    /// If known already, the public key which hashes to `miner_pkh`
+    pub miner_pubkey: Option<StacksPublicKey>,
+    /// the last burn block in this fork which had a sortition
+    pub prior_sortition: ConsensusHash,
+    /// the committed to parent tenure ID
+    pub parent_tenure_id: ConsensusHash,
+    /// this sortition's consensus hash
+    pub consensus_hash: ConsensusHash,
+    /// did the miner in this sortition do something
+    ///  to become invalidated as a miner?
+    pub invalidated: SortitionMinerStatus,
+}
+
+/// The signer's current view of the stacks chain's sortition
+///  state
+pub struct SortitionsView {
+    /// the prior successful sortition (this corresponds to the "prior" miner slot)
+    pub last_sortition: Option<SortitionState>,
+    /// the current successful sortition (this corresponds to the "current" miner slot)
+    pub cur_sortition: Option<SortitionState>,
+    /// is the view fresh?
+    pub fresh: bool,
+    /// the hash at which the sortitions view was last fetched
+    pub latest_consensus_hash: Option<ConsensusHash>,
+}
+
+impl TryFrom<SortitionInfo> for SortitionState {
+    type Error = ClientError;
+    fn try_from(value: SortitionInfo) -> Result<Self, Self::Error> {
+        Ok(Self {
+            miner_pkh: value
+                .miner_pk_hash160
+                .ok_or_else(|| ClientError::UnexpectedSortitionInfo)?,
+            miner_pubkey: None,
+            prior_sortition: value
+                .last_sortition_ch
+                .ok_or_else(|| ClientError::UnexpectedSortitionInfo)?,
+            consensus_hash: value.consensus_hash,
+            parent_tenure_id: value
+                .stacks_parent_ch
+                .ok_or_else(|| ClientError::UnexpectedSortitionInfo)?,
+            invalidated: SortitionMinerStatus::Valid,
+        })
+    }
+}
+
+enum ProposedBy<'a> {
+    LastSortition(&'a SortitionState),
+    CurrentSortition(&'a SortitionState),
+}
+
+impl<'a> ProposedBy<'a> {
+    pub fn state(&self) -> &SortitionState {
+        match self {
+            ProposedBy::LastSortition(ref x) => x,
+            ProposedBy::CurrentSortition(ref x) => x,
+        }
+    }
+}
+
+impl SortitionsView {
+    /// Initialize an empty sortitions view struct -- it will refresh() before
+    ///  checking any proposals.
+    pub fn new() -> Self {
+        Self {
+            last_sortition: None,
+            cur_sortition: None,
+            fresh: false,
+            latest_consensus_hash: None,
+        }
+    }
+
+    /// Apply checks from the SortitionsView on the block proposal.
+    ///
```

Comment:
Specifically, this method seems to be checking that the block proposal is _consistent_ with the `SortitionsView`.  However, there's no guarantee here that the `SortitionsView` itself corresponds to canonical Bitcoin (and thus canonical Stacks) state.  Also, in the event of a fork, which state is assumed to be canonical?  The chainstate in the `signer_db`, or the given block proposal?

Reply:
The response here is related to my other ones above: the sortitions view is simply the latest canonical sortition/burnchain view from the stacks node. The signerdb’s state is all addressed by the tenure IDs, which should be valid irrespective of bitcoin forking.

Reply:
Yes, I see that now.  I was thrown by the fact that this struct persists across multiple calls to `.refresh()`, which I thought was odd since there doesn't appear to be a need for this struct to live beyond checking a block proposal.